[MTIA] Store original file path and line number that FX op derives from. Print the file path and line number for each op in IR printing.

patrick-toulme opened the pull request:
Summary:
When tracing with torch export to FX IR, we need to keep track of the original file path and line number that the FX op originates from in the user code. We then need to print this at the far right of the Node in IR printing.

Without a mapping of user PyTorch code to FX IR, it is very difficult to debug the IR, especially in cases where there are mysterious ops appearing.

I also add support for propagating the line number and file path even with decomposition. Side note: I find it interesting that decomposition actually produces a second traced ExportedProgram. Why do we not just run decomposition during the first tracing?

Jax/XLA currently offers this to users, and it is extremely useful for debug. 

**We default the printing to not print debug file source / line info by default.**

IR Example for Llama: **Scroll to the right**
```
    %div_tensor_12 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.div.Tensor](kwargs = {input: %reshape_default_906, other: 11.313708498984761}) file_path = transformers/models/llama/modeling_llama.py line_number = 419
    %slice_tensor_229 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %clone_default_152, dim: 0, start: 0, end: 1, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_230 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_229, dim: 1, start: 0, end: 1, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_231 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_230, dim: 2, start: 0, end: 2960, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_232 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_231, dim: 3, start: 0, end: 2960, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %add_tensor_135 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.add.Tensor](kwargs = {input: %div_tensor_12, other: %slice_tensor_232, alpha: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %_to_copy_default_194 : f32[1, 32, 2960, 2960][#users=1] = call_function[target=aten._to_copy.default](kwargs = {input: %add_tensor_135, dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: None, non_blocking: False, memory_format: None}) file_path = transformers/models/llama/modeling_llama.py line_number = 426
    %_softmax_default_12 : f32[1, 32, 2960, 2960][#users=1] = call_function[target=aten._softmax.default](kwargs = {input: %_to_copy_default_194, dim: -1, half_to_float: False}) file_path = transformers/models/llama/modeling_llama.py line_number = 426
    %_to_copy_default_195 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten._to_copy.default](kwargs = {input: %_softmax_default_12, dtype: torch.float16, layout: None, device: None, pin_memory: None, non_blocking: False, memory_format: None}) file_path = transformers/models/llama/modeling_llama.py line_number = 426
    %clone_default_74 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.clone.default](kwargs = {input: %_to_copy_default_195, memory_format: None}) file_path = transformers/models/llama/modeling_llama.py line_number = 427
    %repeat_default_173 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.repeat.default](kwargs = {input: %clone_default_74, repeats: [1, 1, 1, 1]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %reshape_default_907 : f16[32, 2960, 2960][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %repeat_default_173, shape: [32, 2960, 2960]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %repeat_default_174 : f16[1, 32, 2960, 128][#users=1] = call_function[target=aten.repeat.default](kwargs = {input: %reshape_default_903, repeats: [1, 1, 1, 1]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %reshape_default_908 : f16[32, 2960, 128][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %repeat_default_174, shape: [32, 2960, 128]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %bmm_default_72 : f16[32, 2960, 128][#users=1] = call_function[target=aten.bmm.default](kwargs = {input: %reshape_default_907, mat2: %reshape_default_908}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %reshape_default_909 : f16[1, 32, 2960, 128][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %bmm_default_72, shape: [1, 32, 2960, 128]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %permute_default_453 : f16[1, 2960, 32, 128][#users=1] = call_function[target=aten.permute.default](kwargs = {input: %reshape_default_909, dims: [0, 2, 1, 3]}) file_path = transformers/models/llama/modeling_llama.py line_number = 436
    %clone_default_75 : f16[1, 2960, 32, 128][#users=1] = call_function[target=aten.clone.default](kwargs = {input: %permute_default_453, memory_format: torch.contiguous_format}) file_path = transformers/models/llama/modeling_llama.py line_number = 436
    %reshape_default_910 : f16[1, 2960, 4096][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %clone_default_75, shape: [1, 2960, -1]}) file_path = transformers/models/llama/modeling_llama.py line_number = 438
    %permute_default_454 : f16[4096, 4096][#users=1] = call_function[target=aten.permute.default](kwargs = {input: %model_model_model_layers_12_self_attn_o_proj_weight, dims: [1, 0]}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %reshape_default_911 : f16[2960, 4096][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %reshape_default_910, shape: [2960, 4096]}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %mm_default_87 : f16[2960, 4096][#users=1] = call_function[target=aten.mm.default](kwargs = {input: %reshape_default_911, mat2: %permute_default_454}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %reshape_default_912 : f16[1, 2960, 4096][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %mm_default_87, shape: [1, 2960, 4096]}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %add_tensor_136 : f16[1, 2960, 4096][#users=2] = call_function[target=aten.add.Tensor](kwargs = {input: %add_tensor_131, other: %reshape_default_912, alpha: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 740
```

Test Plan: Added unit test

Differential Revision: D73346635




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

zou3519 commented on torch/export/exported_program.py:
```diff
@@ -1,5 +1,6 @@
 # mypy: allow-untyped-decorators
 # mypy: allow-untyped-defs
+import collections
```
Have you tried gm.print_readable() ? That should have line number information (it does for compile, I'm not sure about export)

patrick-toulme replied:
I just tried it on Llama. It does print source locations, but many are not in the user scope. For example, it maps many ops to a location in PyTorch, but the user code that calls that PyTorch location is a few frames up in the stack trace. 

MTIA also uses a custom IR printer, so we need the source location and file info saved per node, so we can manipulate it in our custom IR printer. I have made this change behind a flag now defaulted to False.

patrick-toulme replied:
MTIA also needs us for data dependent dynamic shape debug.

angelayi replied:
I think print_readable() might only print the bottom-most frame for the stacktrace. I wonder if the actual stacktrace on node.meta["stack_trace"] contains all the lines that you need? Then we can just rely on that value.

patrick-toulme replied:
stack_trace is not always present in Node meta. There is also the other issue, that if we use stack_trace we would have to parse the file_path and line number anyway. MTIA dumps the IR after every pass, so parsing the file path and line number from every node's stack trace in each dump is not really practical, especially for graphs with 100k nodes.

There is also the other issue that stack_trace is wrong after decomposition. After running decomposition during traces the stack_trace is from the decomposition methods, not the original PyTorch code.

I do think this is change is very useful for PyTorch users, especially those users who want to see how their FX ops map to PyTorch code.

angelayi replied:
> stack_trace is not always present in Node meta. 

If you are directly exporting the model with torch.export there should be stacktrace on every node, otherwise it's a bug. But iirc MTIA first goes through a different tracer so it might be helpful to check if that tracer is correctly setting the flag to capture [stacktraces](https://github.com/pytorch/pytorch/blob/main/torch/fx/proxy.py#L126). 

> if we use stack_trace we would have to parse the file_path and line number anyway

stack_trace should just be a string which you could dump before every node? 

Here's an example of the stacktraces:
```python
lass M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(3, 3)

    def forward(self, x):
        x = x + x 
        res = self.linear(x)
        return res

ep = torch.export.export(M(), (torch.ones(3, 3),))
for node in ep.graph.nodes:
    print(node, node.meta.get("stack_trace"))
"""
p_linear_weight None
p_linear_bias None
x None
add File "/data/users/angelayi/pytorch/moo.py", line 3143, in forward
    x = x + x
linear File "/data/users/angelayi/pytorch/moo.py", line 3144, in forward
    res = self.linear(x)
File "/data/users/angelayi/pytorch/torch/nn/modules/linear.py", line 125, in forward
    return F.linear(input, self.weight, self.bias)
output None
"""
ep = ep.run_decompositions(None)
for node in ep.graph.nodes:
    print(node, node.meta.get("stack_trace"))
"""
p_linear_weight None
p_linear_bias None
x None
add File "/data/users/angelayi/pytorch/moo.py", line 3143, in forward
    x = x + x
permute File "/data/users/angelayi/pytorch/moo.py", line 3144, in forward
    res = self.linear(x)
File "/data/users/angelayi/pytorch/torch/nn/modules/linear.py", line 125, in forward
    return F.linear(input, self.weight, self.bias)
addmm File "/data/users/angelayi/pytorch/moo.py", line 3144, in forward
    res = self.linear(x)
File "/data/users/angelayi/pytorch/torch/nn/modules/linear.py", line 125, in forward
    return F.linear(input, self.weight, self.bias)
output None
"""
```

angelayi replied:
Maybe the better fix is to filter out the stacktraces better so that instead of seeing `return F.linear(input, self.weight, self.bias)` we directly see `res = self.linear(x)`

patrick-toulme replied:
For MTIA, when I enable stack_trace by manually enabling it I get stack traces that are dozens of frames deep per node. After custom MTIA decomposition, stack_trace is also changed due to retracing.

MTIA is using export. For some of our graphs, we have 100k+ FX Nodes. We also want to dump the IR after all our passes, which is a very large amount of passes. We cannot parse stack_trace each time.

This change would be very beneficial for MTIA compiler development :)

patrick-toulme replied:
> Maybe the better fix is to filter out the stacktraces better so that instead of seeing `return F.linear(input, self.weight, self.bias)` we directly see `res = self.linear(x)`

I can see a use case where we ideally want the whole stack_trace. IMO stack_trace, file_path and line_number are all different. 

For context, Jax/XLA allows user to get the file path and line number next to the FX op if XLA debug flag is enabled. 

Note MTIA dumps IR differently than normal torch graph module printing. MTIA dumps IR closer to XLA style so path and line at right side of op 

```
    %div_tensor_12 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.div.Tensor](kwargs = {input: %reshape_default_906, other: 11.313708498984761}) file_path = transformers/models/llama/modeling_llama.py line_number = 419
    %slice_tensor_229 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %clone_default_152, dim: 0, start: 0, end: 1, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_230 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_229, dim: 1, start: 0, end: 1, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_231 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_230, dim: 2, start: 0, end: 2960, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_232 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_231, dim: 3, start: 0, end: 2960, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
```

jfix71 replied:
Hey @patrick-toulme thanks for working on this. Some questions:

> For MTIA, when I enable stack_trace by manually enabling it I get stack traces that are dozens of frames deep per node.

This does seem like overkill -- but just to be sure, it contains the info you want, right?

> After custom MTIA decomposition, stack_trace is also changed due to retracing. 

I assume this referring to calling `ep.run_decompositions`? If so, and assuming you're saying the original stack trace is lost, this seems like a bug that we should file with the export team.

> MTIA is using export. For some of our graphs, we have 100k+ FX Nodes. We also want to dump the IR after all our passes, which is a very large amount of passes. We cannot parse stack_trace each time.

Agreed, but wondering if we could just extract from stack trace the file path + number once post initial export. Then we achieve the same but without needing to add more functionality to export.

patrick-toulme replied:
On today's warm stack_trace before and after decomposition is correct. I am not sure why on Monday when I started working on this the stack_trace after decomposition was different. Was anything changed in the past two days?

zou3519 reviewed (changes requested):
tests failing

patrick-toulme committed:
[MTIA] Store original file path and line number that FX op derives from. Print the file path and line number for each op in IR printing. (#151851)

Summary:

When tracing with torch export to FX IR, we need to keep track of the original file path and line number that the FX op originates from in the user code. We then need to print this at the far right of the Node in IR printing.

Without a mapping of user PyTorch code to FX IR, it is very difficult to debug the IR, especially in cases where there are mysterious ops appearing.

I also add support for propagating the line number and file path even with decomposition. Side note: I find it interesting that decomposition actually produces a second traced ExportedProgram. Why do we not just run decomposition during the first tracing?

Jax/XLA currently offers this to users, and it is extremely useful for debug. 

**We default the printing to not print debug file source / line info by default.**

Controlled by env variable TORCH_FX_CAPTURE_SOURCE_LOCATIONS

IR Example for Llama: **Scroll to the right**
```
    %div_tensor_12 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.div.Tensor](kwargs = {input: %reshape_default_906, other: 11.313708498984761}) file_path = transformers/models/llama/modeling_llama.py line_number = 419
    %slice_tensor_229 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %clone_default_152, dim: 0, start: 0, end: 1, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_230 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_229, dim: 1, start: 0, end: 1, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_231 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_230, dim: 2, start: 0, end: 2960, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %slice_tensor_232 : f16[1, 1, 2960, 2960][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor_231, dim: 3, start: 0, end: 2960, step: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %add_tensor_135 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.add.Tensor](kwargs = {input: %div_tensor_12, other: %slice_tensor_232, alpha: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 423
    %_to_copy_default_194 : f32[1, 32, 2960, 2960][#users=1] = call_function[target=aten._to_copy.default](kwargs = {input: %add_tensor_135, dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: None, non_blocking: False, memory_format: None}) file_path = transformers/models/llama/modeling_llama.py line_number = 426
    %_softmax_default_12 : f32[1, 32, 2960, 2960][#users=1] = call_function[target=aten._softmax.default](kwargs = {input: %_to_copy_default_194, dim: -1, half_to_float: False}) file_path = transformers/models/llama/modeling_llama.py line_number = 426
    %_to_copy_default_195 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten._to_copy.default](kwargs = {input: %_softmax_default_12, dtype: torch.float16, layout: None, device: None, pin_memory: None, non_blocking: False, memory_format: None}) file_path = transformers/models/llama/modeling_llama.py line_number = 426
    %clone_default_74 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.clone.default](kwargs = {input: %_to_copy_default_195, memory_format: None}) file_path = transformers/models/llama/modeling_llama.py line_number = 427
    %repeat_default_173 : f16[1, 32, 2960, 2960][#users=1] = call_function[target=aten.repeat.default](kwargs = {input: %clone_default_74, repeats: [1, 1, 1, 1]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %reshape_default_907 : f16[32, 2960, 2960][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %repeat_default_173, shape: [32, 2960, 2960]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %repeat_default_174 : f16[1, 32, 2960, 128][#users=1] = call_function[target=aten.repeat.default](kwargs = {input: %reshape_default_903, repeats: [1, 1, 1, 1]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %reshape_default_908 : f16[32, 2960, 128][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %repeat_default_174, shape: [32, 2960, 128]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %bmm_default_72 : f16[32, 2960, 128][#users=1] = call_function[target=aten.bmm.default](kwargs = {input: %reshape_default_907, mat2: %reshape_default_908}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %reshape_default_909 : f16[1, 32, 2960, 128][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %bmm_default_72, shape: [1, 32, 2960, 128]}) file_path = transformers/models/llama/modeling_llama.py line_number = 428
    %permute_default_453 : f16[1, 2960, 32, 128][#users=1] = call_function[target=aten.permute.default](kwargs = {input: %reshape_default_909, dims: [0, 2, 1, 3]}) file_path = transformers/models/llama/modeling_llama.py line_number = 436
    %clone_default_75 : f16[1, 2960, 32, 128][#users=1] = call_function[target=aten.clone.default](kwargs = {input: %permute_default_453, memory_format: torch.contiguous_format}) file_path = transformers/models/llama/modeling_llama.py line_number = 436
    %reshape_default_910 : f16[1, 2960, 4096][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %clone_default_75, shape: [1, 2960, -1]}) file_path = transformers/models/llama/modeling_llama.py line_number = 438
    %permute_default_454 : f16[4096, 4096][#users=1] = call_function[target=aten.permute.default](kwargs = {input: %model_model_model_layers_12_self_attn_o_proj_weight, dims: [1, 0]}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %reshape_default_911 : f16[2960, 4096][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %reshape_default_910, shape: [2960, 4096]}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %mm_default_87 : f16[2960, 4096][#users=1] = call_function[target=aten.mm.default](kwargs = {input: %reshape_default_911, mat2: %permute_default_454}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %reshape_default_912 : f16[1, 2960, 4096][#users=1] = call_function[target=aten.reshape.default](kwargs = {input: %mm_default_87, shape: [1, 2960, 4096]}) file_path = transformers/models/llama/modeling_llama.py line_number = 445
    %add_tensor_136 : f16[1, 2960, 4096][#users=2] = call_function[target=aten.add.Tensor](kwargs = {input: %add_tensor_131, other: %reshape_default_912, alpha: 1}) file_path = transformers/models/llama/modeling_llama.py line_number = 740
```

Test Plan: Added unit test

Differential Revision: D73346635

patrick-toulme commented:
> tests failing

@zou3519 I resolved the test issues. One test is failing due to invalid AWS credentials. Thank you!

patrick-toulme commented on torch/export/exported_program.py:
```diff
@@ -816,16 +817,25 @@ def _common_getitem_elimination_pass(
+    if torch.fx.config.capture_source_locations:
+        # Update node line_number and file_path.
+        for old_node in old_gm.graph.nodes:
+            for node in old_node_to_new_node[old_node.name]:
```
@angelayi  This fixes the decomposition.
